sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/UnsafeArraySuite.scala

```diff
@@ -18,27 +18,187 @@
+    assert(unsafeInt.isInstanceOf[UnsafeArrayData])
+    assert(unsafeInt.numElements == intArray.length)
+    intArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeInt.getInt(i) == e)
+    }
+
+    val unsafeLong = ExpressionEncoder[Array[Long]].resolveAndBind().
+      toRow(longArray).getArray(0)
+    assert(unsafeLong.isInstanceOf[UnsafeArrayData])
+    assert(unsafeLong.numElements == longArray.length)
+    longArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeLong.getLong(i) == e)
+    }
+
+    val unsafeFloat = ExpressionEncoder[Array[Float]].resolveAndBind().
+      toRow(floatArray).getArray(0)
+    assert(unsafeFloat.isInstanceOf[UnsafeArrayData])
+    assert(unsafeFloat.numElements == floatArray.length)
+    floatArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeFloat.getFloat(i) == e)
+    }
+
+    val unsafeDouble = ExpressionEncoder[Array[Double]].resolveAndBind().
+      toRow(doubleArray).getArray(0)
+    assert(unsafeDouble.isInstanceOf[UnsafeArrayData])
+    assert(unsafeDouble.numElements == doubleArray.length)
+    doubleArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeDouble.getDouble(i) == e)
+    }
+
+    val unsafeString = ExpressionEncoder[Array[String]].resolveAndBind().
+      toRow(stringArray).getArray(0)
+    assert(unsafeString.isInstanceOf[UnsafeArrayData])
+    assert(unsafeString.numElements == stringArray.length)
+    stringArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeString.getUTF8String(i).toString().equals(e))
+    }
+
+    val unsafeDate = ExpressionEncoder[Array[Int]].resolveAndBind().
+      toRow(dateArray).getArray(0)
+    assert(unsafeDate.isInstanceOf[UnsafeArrayData])
+    assert(unsafeDate.numElements == dateArray.length)
+    dateArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeDate.get(i, DateType) == e)
+    }
+
+    val unsafeTimestamp = ExpressionEncoder[Array[Long]].resolveAndBind().
+      toRow(timestampArray).getArray(0)
+    assert(unsafeTimestamp.isInstanceOf[UnsafeArrayData])
+    assert(unsafeTimestamp.numElements == timestampArray.length)
+    timestampArray.zipWithIndex.map { case (e, i) =>
+      assert(unsafeTimestamp.get(i, TimestampType) == e)
+    }
+
+    Seq(decimalArray4_1, decimalArray20_20).map { decimalArray =>
+      val decimal = decimalArray(0)
+      val schema = new StructType().add(
+        "array", ArrayType(DecimalType(decimal.precision, decimal.scale)))
+      val encoder = RowEncoder(schema).resolveAndBind()
+      val externalRow = Row(decimalArray)
+      val ir = encoder.toRow(externalRow)
+
+      val unsafeDecimal = ir.getArray(0)
+      assert(unsafeDecimal.isInstanceOf[UnsafeArrayData])
+      assert(unsafeDecimal.numElements == decimalArray.length)
+      decimalArray.zipWithIndex.map { case (e, i) =>
+        assert(unsafeDecimal.getDecimal(i, e.precision, e.scale).toBigDecimal == e)
+      }
+    }
+
+    val unsafeInterval = ExpressionEncoder[Array[CalendarInterval]].resolveAndBind().
```

Comment:
`CalendarInterval` is not external, let's follow the way we test decimal, i.e. use `RowEncoder`